hawc/apps/epi/api.py

```diff
@@ -87,37 +99,410 @@ def result_heatmap(self, request, pk):
+        self.handle_criteria(serializer.instance)
+        self.process_criteria_association(serializer, serializer.instance.id, True)
+
+    def perform_update(self, serializer):
+        super().perform_update(serializer)
+        self.handle_criteria(serializer.instance)
+        self.process_criteria_association(serializer, serializer.instance.id, False)
+
+    @transaction.atomic
+    def update(self, request, *args, **kwargs):
+        return super().update(request, *args, **kwargs)
+
+    @transaction.atomic
+    def create(self, request, *args, **kwargs):
+        # Default behavior except we need to refresh the serializer to get the criteria to show up in the return...
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        self.perform_create(serializer)
+
+        instance = self.model.objects.get(id=serializer.data["id"])
+        serializer = self.get_serializer(instance)
+        return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+
+class Exposure(ReadWriteSerializerMixin, PermCheckerMixin, AssessmentEditViewset):
+    perm_checker_key = "study_population"
     assessment_filter_args = "study_population__study__assessment"
     model = models.Exposure
-    serializer_class = serializers.ExposureSerializer
+    read_serializer_class = serializers.ExposureSerializer
+    write_serializer_class = serializers.ExposureWriteSerializer
+
+    def handle_dtxsid(self, request):
+        """
+        Checks supplied dtxsid if present for existence in the database; attempting to create on-the-fly if necessary.
+        """
+        if "dtxsid" in request.data:
+            dtxsid_probe = request.data["dtxsid"]
+            try:
+                DSSTox.objects.get(dtxsid=dtxsid_probe)
+            except ObjectDoesNotExist:
+                try:
+                    substance = DssSubstance.create_from_dtxsid(dtxsid_probe)
+
+                    dsstox = DSSTox(dtxsid=substance.dtxsid, content=substance.content)
+                    dsstox.save()
+                except ValueError:
+                    raise ValidationError(
+                        f"dtxsid '{dtxsid_probe}' does not exist and could not be imported"
+                    )
+
+    @transaction.atomic
+    def update(self, request, *args, **kwargs):
+        self.handle_dtxsid(request)
+
+        return super().update(request, *args, **kwargs)
+
+    @transaction.atomic
+    def create(self, request, *args, **kwargs):
+        self.handle_dtxsid(request)
+
+        # Default behavior except we need to refresh the serializer to get the central tendencies to show up in the return...
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        self.perform_create(serializer)
+
+        # Refresh serializer instance, as we want to get the central tendencies we created...
+        instance = self.model.objects.get(id=serializer.data["id"])
```

Comment:
can you just do `serializer.instance.refresh_from_db()` here and elsewhere?

Reply:
done; I think I modeled this after what the animal api is doing as it uses similar calls. Swithed to refersh_from_db.